### ISSUE TITLE
Explain how to interpret output of .predict() in API doc

### ISSUE THREAD

**Issue by MaxPowerWasTaken** [sebp/scikit-survival#15]
(I also posted this as a question on Stack Overflow: https://stackoverflow.com/q/47274356/1870832 )

I'm confused how to interpret the output of .predict from a fitted CoxnetSurvivalAnalysis model in scikit-survival. I've read through the notebook Intro to Survival Analysis in scikit-survival and the API reference, but can't find an explanation. Below is a minimal example of what leads to my confusion:

```
import pandas as pd
from sksurv.datasets import load_veterans_lung_cancer
from sksurv.linear_model import CoxnetSurvivalAnalysis

# load data
data_X, data_y = load_veterans_lung_cancer()

# one-hot-encode categorical columns in X
categorical_cols = ['Celltype', 'Prior_therapy', 'Treatment']

X = data_X.copy()
for c in categorical_cols:
    dummy_matrix = pd.get_dummies(X[c], prefix=c, drop_first=False)
    X = pd.concat([X, dummy_matrix], axis=1).drop(c, axis=1)

# display final X to fit Cox Elastic Net model on
del data_X
print(X.head(3))

```
so here's the X going into the model:

```
   Age_in_years  Celltype  Karnofsky_score  Months_from_Diagnosis  \
0          69.0  squamous             60.0                    7.0   
1          64.0  squamous             70.0                    5.0   
2          38.0  squamous             60.0                    3.0   

  Prior_therapy Treatment  
0            no  standard  
1           yes  standard  
2            no  standard  

```
...moving on to fitting model and generating predictions:

```
# Fit Model
coxnet_model = CoxnetSurvivalAnalysis()
coxnet.fit(X, data_y)    

# What are these predictions?    
preds = coxnet.predict(X)

```
preds has same number of records as X, but their values are wayyy different than the values in data_y, even when predicted on the same data they were fit on.

```
print(preds.mean()) 
print(data_y['Survival_in_days'].mean())
```
output:

```
-0.044114643249153422
121.62773722627738

```
So what exactly are preds? Clearly .predict means something pretty different here than in scikit-learn, but I can't figure out what. The API Reference says it returns "The predicted decision function," but what does that mean? And how do I get to the predicted estimate in months yhat for a given X? I'm new to survival analysis so I'm obviously missing something.

**Comment by pavopax** [comment]
AFAIK:

It returns a type of risk score. Higher value means  higher risk of your event (class value = True). 

Recall that with censored data, your `y` is not just a continuous value (time) but also an indicator (censor) and risk is a summary of these two components. 

You were probably looking for a predicted time. You can get the predicted survival function with `estimator.predict_survival_function` as in the example 00 notebook. 

From that, an alternative prediction could be a median predicted survival time. 

(Actually, I’m trying to extract this but it’s been a bit of a pain to munge. @sebp do you have thoughts on this?)

**Comment by sebp** [comment]
It seems the API documentation needs to be improved to explain how to interpret predictions.

The semantics of predictions in survival models are very different from traditional machine learning models, such as those from scikit-learn. As @pavopax explained, predictions are risk scores on an arbitrary scale, which means you can usually only determine the sequence of events, but not their exact time.

The main reason for this is that you would need to specify a parametric distribution of the survival function when estimating the probability of survival at a given time. This step is not necessary when using Cox's proportional hazards, because of its proportional hazards assumption. You can use [ predict_survival_function](https://scikit-survival.readthedocs.io/en/latest/generated/sksurv.linear_model.CoxPHSurvivalAnalysis.html#sksurv.linear_model.CoxPHSurvivalAnalysis.predict_survival_function) or [predict_cumulative_hazard_function](https://scikit-survival.readthedocs.io/en/latest/generated/sksurv.linear_model.CoxPHSurvivalAnalysis.html#sksurv.linear_model.CoxPHSurvivalAnalysis.predict_cumulative_hazard_function).

Median survival time has essentially the same problem, because it is defined as the 50th percentile of a distribution, which needs to be specified.

If you are using a model other than Cox's model and still want to get a predicted survival curve, a work-around would be to use the predicted risk score as a feature in Cox's model and get a predicted survival curve this way. Obviously, you can't avoid the proportional hazards assumption this way. Other ways are possible, e.g. [Van Belle et al.](http://www.jmlr.org/papers/v12/vanbelle11a.html) fit a monotone least squares support vector regression model to estimate the survival function from predicted scores.

**Comment by sebp** [comment]
I added [more explanations](https://github.com/sebp/scikit-survival/blob/0768703f8cad1aeb846477235edcc0124c7dbd28/doc/understanding_predictions.rst) to the documentation, which hopefully clarifies your questions. Please have a look and let me know if parts are unclear or important information is missing.

**Comment by MaxPowerWasTaken** [comment]
Hey thanks a lot sebp. I'll look over this today.

(Just a heads up, looks like some of the mathjax didn't render as intended)

**Comment by olivierzach** [comment]
Quick question regarding the predicted rank order - would the min value of the predict output be the highest risk to event? Docs mention ascending order as the predicted order of events but want to make sure this is the case.

**Comment by pavopax** [comment]
My understanding is: higher value of risk score means higher risk of your event (class value = 1 or True).

In which case, it would be the opposite: min value of predict output means _lowest_ risk of event.

**Comment by CamilaFernandez8** [comment]
Hello,
I think I have kind of the same issue (not understanding how to get the survival function). I tried to follow this talk from @sebp 

https://www.youtube.com/watch?v=hdvbP6aJRCw&t=751s 

to use and plot the predict_survival_function but he didn't specify what to use as the input in DataFrame, so I guess it should be a matrix with the feature of the n samples, but if I do that I get n different curves and I don't know how to get the survival function. Should it be the mean of that curves?

Also, I am trying to get the survival function from a pipeline using make_pipeline(preprocessor,CoxPHSurvivalAnalysis) but pipeline doesn't have 'predict_survival_function' as an attribute so I only can get the risks with %.predict. How could I get the survival function using the pipeline?

I am also new on this :( Thank you a lot!

Cam

**Comment by sebp** [comment]
Yes, the input to `predict_survival_function` should be an array with shape `(n_samples, n_features)` and will return one survival function for each array. An example is available in the [introduction notebook](https://nbviewer.jupyter.org/github/sebp/scikit-survival/blob/master/examples/00-introduction.ipynb#Multivariate-Survival-Models).

Unfortunately, scikit-learn's Pipeline is not aware of `predict_survival_function`. There might be a way to patch this, but for now you would need to get the survival estimator from the pipeline and directly call `predict_survival_function`:
```python
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sksurv.linear_models import CoxPHSurvivalAnalysis

pipe = make_pipeline(StandardScaler(), CoxPHSurvivalAnalysis())
pipe.fit(X, y)
survival_model = pipe.steps[-1][1]
surv_fns = survival_model.predict_survival_function(X_predict)
```

**Comment by hermidalc** [comment]
> Yes, the input to `predict_survival_function` should be an array with shape `(n_samples, n_features)` and will return one survival function for each array. An example is available in the [introduction notebook](https://nbviewer.jupyter.org/github/sebp/scikit-survival/blob/master/examples/00-introduction.ipynb#Multivariate-Survival-Models).
> 
> Unfortunately, scikit-learn's Pipeline is not aware of `predict_survival_function`. There might be a way to patch this, but for now you would need to get the survival estimator from the pipeline and directly call `predict_survival_function`:
> 
> ```python
> from sklearn.pipeline import make_pipeline
> from sklearn.preprocessing import StandardScaler
> from sksurv.linear_models import CoxPHSurvivalAnalysis
> 
> pipe = make_pipeline(StandardScaler(), CoxPHSurvivalAnalysis())
> pipe.fit(X, y)
> survival_model = pipe.steps[-1][1]
> surv_fns = survival_model.predict_survival_function(X_predict)
> ```

Hi @sebp - one issue I’m trying to understand regarding `predict_survival_function`, `predict`, and `predict_cumulative_hazard_function` in both `CoxnetSurvivalAnalysis` and `CoxPHSurvivalAnalysis` is whether `X_predict` should be standardized or normalized before passing it to these methods if the training data originally used to `fit` the model was standardized or normalized.

In your example above you don’t do that. Also, in the source code for `CoxnetSurvivalAnalysis` if you set `normalize=True` I see it’s only normalizing the training data during `fit` and not any test data passed to `predict`, `predict_survival_function`, or `predict_cumulative_hazard_function`.

**Comment by amaliaspa** [comment]
Hi,
I am trying to use random survival forests to predict remaining median survival time for the censored observations. If I understand well from the thread above, the idea is to call `predict_survival_function ` and then extract from it the median survival time for a specific observation. My question is - is this time considered since the very beginning (time 0), or the remaining time (i.e. the algorithm take into account that one specific observation already lived up to time t )? Otherwise said, should I subtract time t from the median predicted survival time to get an estimate about remaining time? Please let me know if you have another suggestion. Thanks!

**Comment by sebp** [comment]
predict_survival_function(X) returns the survival function S(t|X), i.e. 
the probability of survival beyond time t, P(T > t | X). The median 
survival time would be the time point t for which S(t|X) = 0.5. If you 
need the survival function conditional on survival at least to t_0, have 
a look at issue #128

**Comment by epx-dan** [comment]
**_TL;DR_:**
The output of `.predict()` is the exponent <img src="https://render.githubusercontent.com/render/math?math={X_i^T\beta}"> based on CoxPH formula shown below.

==========

I've tried to manually do the calculation and I think I've figured out what `CoxPHSurvivalAnalysis().predict()` is returning, and here is my understanding:

```python
# Assume there are 10 features:
X_i = X_test.iloc[0, :] # (10,) array, indicating i-th patient's covariates
coxph = CoxPHSurvivalAnalysis()
preds = coxph.predict(x_i) # (1,) array, e.g., array([-1.81138421])
betas = coxph.coef_ # this is the coefficient vector common to *all* observations
sv_function_i = coxph.predict_survival_function() # survival function for i-th patient, e.g., array([StepFunction(x=array([  15.,   16.,   17., ..., 1586., 1591., 1623.]), y=array([9.99956830e-01, 9.99913136e-01, 9.92522412e-01, ..., 3.19995148e-06, 1.98618910e-06, 6.85570033e-07]), a=1.0, b=0.0)], dtype=object)
```

**_Here is the math part:_**
Based on CoxPH formula: <img src="https://render.githubusercontent.com/render/math?math=h_i(t)=\lambda(t)e^{X_i^T\beta}">, ℎ𝑖(𝑡) is the i-th patient's hazard function, 𝜆(𝑡) is the base hazard function, exp(𝐱^𝑇_𝑖𝛽) is the **risk score** for the i-th patient. According to the `get_survival_function(...)` under `class BreslowEstimator` from `coxph.py`, the i-th patient's survival function is calculated by `np.power(baseline_survival, risk_score_i)` where baseline_survival [𝜆(𝑡)] is returned by `coxph.baseline_survival_` and risk_score_i [exp(𝐱^𝑇_𝑖𝛽)] can be calculated by `np.exp( np.sum(patient_ith_X * coxph.coef_) )` or `np.exp( np.dot(patient_ith_X, coxph.coef_) )`.

```python
# Calculate the result returned from ".predict_survival_function()" step by step would be like:
coxph_base_sv_function = coxph.baseline_survival_.y # this is 𝜆(𝑡), the base survival function common to all patients, e.g., y=array([9.99956830e-01, 9.99913136e-01, 9.92522412e-01, ..., 3.19995148e-06, 1.98618910e-06, 6.85570033e-07])
coxph_risk_score_i = np.exp(preds) # remember ".predict()" is the exponent 𝐱^𝑇_𝑖𝛽, so then take the exponential of it will get the i-th patient's risk score which is a numeric value, e.g., 0.16342776226883848
coxph_sv_function_i = np.power(coxph_base_sv_function, coxph_risk_score_i) # this is exactly the same with sv_function_i
```

**_AND HERE IS THE ESSENSE:_**
The result from `preds`, a single numeric value, is equal to `np.dot(patient_ith_X, coxph.coef_)` which is <img src="https://render.githubusercontent.com/render/math?math={X_i^T\beta}">.

**_Takeaway:_**
If want to get the **risk score** for i-th patient, just calculate the exponential of `.predict()`, e.g., `np.exp(coxph.predict(X_i))` which is <img src="https://render.githubusercontent.com/render/math?math=e^{X_i^T\beta}">.

**Comment by M4ieni** [comment]
> predict_survival_function(X) returns the survival function S(t|X), i.e. the probability of survival beyond time t, P(T > t | X). The median survival time would be the time point t for which S(t|X) = 0.5. If you need the survival function conditional on survival at least to t_0, have a look at issue #128

Sorry for that question, but how to get the values for  S(t|X) = 0.5 from stepfunction? I want to get t, where the survival function  reached 0.5. thank you :)

**Comment by sebp** [comment]
You need to compute the inverse survival function:
```python
from scipy.interpolate import interp1d

times, probs = kaplan_meier_estimator(y_event, y_time)
# inverse survival function
inverse_km = interp1d(probs[::-1], times[::-1], bounds_error=True, assume_sorted=True)
# S(t) = 0.5 <=> t = S^{-1}(0.5) = t
t = inverse_km(0.5)
```

**Comment by jackyko1991** [comment]
> ```python
> kaplan_meier_estimator
> ```

@sebp after getting individual suvival function using `predict`, why we need to use `kaplan_meier_estimator` to get the median suvival time? 

Also if the survival rate never lower than 0.5 in the training data, will it be still reliable to use median survival time?

**Comment by sebp** [comment]
@jackyko1991 You don't need `kaplan_meier_estimator`, this was just an example. You can also invert the survival function returned by `predict` to get the median survival function.

**Comment by ironb25** [comment]
@sebp I am new to survival analysis. Whats the best strategy to divide subjects into high and low risk after getting risk scores ? Any suggestions? I am using this to get the risk scores:`np.exp(coxph.predict(X_i))` as explained by @epx-dan  (https://github.com/sebp/scikit-survival/issues/15#issuecomment-665756105). And once we divide it, we can do Kaplan-Meier curves to see if the groups are significant? Any way to do this in scikit-survival?

**Comment by pavopax** [comment]
> @sebp I am new to survival analysis. Whats the best strategy to divide subjects into high and low risk after getting risk scores ? Any suggestions? I am using this to get the risk scores:`np.exp(coxph.predict(X_i))` as explained by @epx-dan ([#15 (comment)](https://github.com/sebp/scikit-survival/issues/15#issuecomment-665756105)). And once we divide it, we can do Kaplan-Meier curves to see if the groups are significant? Any way to do this in scikit-survival?

Briefly, common cutoffs for risk are by median (high/low risk) or possibly by top 25% vs bottom 25%. 

To check for significance, I'd recommend first withholding, say, 20% of all of your data, and then do all model development on the 80%. Once you have your model, then "predict" and check the risk groups on the remaining 20%.

**Comment by ironb25** [comment]
Thanks @pavopax So we do Kaplan Meier curves stratifying high vs low risk, only on the 20% test set showing significance(if found)?

**Comment by ironb25** [comment]
Okay, Thank you so much for quick assistance! @pavopax
